File: surf-api-bukkit/surf-api-bukkit-server/src/main/java/dev/slne/surf/surfapi/bukkit/server/reflection/Reflection.java
package dev.slne.surf.surfapi.bukkit.server.reflection;

import dev.slne.surf.surfapi.core.api.reflection.SurfReflection;
import dev.slne.surf.surfapi.core.server.impl.reflection.SurfReflectionImpl;
import xyz.jpenilla.reflectionremapper.ReflectionRemapper;
import xyz.jpenilla.reflectionremapper.proxy.ReflectionProxyFactory;

public final class Reflection {

  public static final ServerStatsCounterProxy SERVER_STATS_COUNTER_PROXY;
  public static final SynchedEntityDataProxy SYNCHED_ENTITY_DATA_PROXY;
  public static final ItemProxy ITEM_PROXY;

  static {
    final ReflectionRemapper remapper = ReflectionRemapper.forReobfMappingsInPaperJar();
    final ReflectionProxyFactory proxyFactory = ReflectionProxyFactory.create(remapper,
        Reflection.class.getClassLoader());
    final SurfReflection surfReflection = SurfReflectionImpl.INSTANCE;

    SERVER_STATS_COUNTER_PROXY = proxyFactory.reflectionProxy(ServerStatsCounterProxy.class);
    SYNCHED_ENTITY_DATA_PROXY = proxyFactory.reflectionProxy(SynchedEntityDataProxy.class);
    ITEM_PROXY = surfReflection.createProxy(ItemProxy.class);

    // gc the remapper
    System.gc();
  }
}

File: surf-api-bukkit/surf-api-bukkit-server/src/main/kotlin/dev/slne/surf/surfapi/bukkit/server/reflection/Reflection.kt
package dev.slne.surf.surfapi.bukkit.server.reflection

import dev.slne.surf.surfapi.bukkit.api.util.reflectionProxy
import dev.slne.surf.surfapi.core.api.reflection.createProxy
import dev.slne.surf.surfapi.core.api.reflection.surfReflection
import xyz.jpenilla.reflectionremapper.ReflectionRemapper
import xyz.jpenilla.reflectionremapper.proxy.ReflectionProxyFactory

object Reflection {
    val SERVER_STATS_COUNTER_PROXY: ServerStatsCounterProxy
    val ITEM_PROXY: ItemProxy
    val ENTITY_PROXY: EntityProxy
    val SERVER_CONNECTION_LISTENER_PROXY: ServerConnectionListenerProxy
    val JAVA_PLUGIN_PROXY: JavaPluginProxy

    init {
        val remapper = ReflectionRemapper.forReobfMappingsInPaperJar()
        val proxyFactory =
            ReflectionProxyFactory.create(remapper, Reflection::class.java.classLoader)

        SERVER_STATS_COUNTER_PROXY = proxyFactory.reflectionProxy<ServerStatsCounterProxy>()
        ITEM_PROXY = surfReflection.createProxy<ItemProxy>()
        ENTITY_PROXY = proxyFactory.reflectionProxy<EntityProxy>()
        SERVER_CONNECTION_LISTENER_PROXY = proxyFactory.reflectionProxy<ServerConnectionListenerProxy>()
        JAVA_PLUGIN_PROXY = surfReflection.createProxy<JavaPluginProxy>()

        // gc the remapper
        System.gc()
    }
}
